step-functions-plugins/step-functions-plugins-java/step-functions-plugins-java-def/src/main/java/step/plugins/java/AbstractScriptFunctionType.java

```diff
@@ -59,41 +57,48 @@ public AbstractScriptFunctionType(Configuration configuration) {
 	@Override
 	public void init() {
 		super.init();
-		daemonJar = registerResource(getClass().getClassLoader(), "java-plugin-handler.jar", false);
+		handlerPackageVersion = registerResource(getClass().getClassLoader(), "java-plugin-handler.jar", false, false);
 	}
 
 	@Override
-	public Map<String, String> getHandlerProperties(T function, AbstractStepContext executionContext) {
+	public HandlerProperties getHandlerProperties(T function, AbstractStepContext executionContext) {
 		Map<String, String> props = new HashMap<>();
 		props.put(ScriptHandler.SCRIPT_LANGUAGE, function.getScriptLanguage().get());
-		registerFile(function.getLibrariesFile(), ScriptHandler.LIBRARIES_FILE, props, true, executionContext);
-		addPluginLibsIfRequired(function.getScriptLanguage().get(), props);
-		registerFile(function.getScriptFile(), ScriptHandler.SCRIPT_FILE, props, true, executionContext);
-		registerFile(function.getErrorHandlerFile(), ScriptHandler.ERROR_HANDLER_FILE, props, true, executionContext);
+		FileVersion libFileVersion = registerFile(function.getLibrariesFile(), ScriptHandler.LIBRARIES_FILE, props, true, executionContext);
+		FileVersion pluginLibFileVersion = addPluginLibsIfRequired(function.getScriptLanguage().get(), props);
+		FileVersion scriptFileVersion = registerFile(function.getScriptFile(), ScriptHandler.SCRIPT_FILE, props, true, executionContext);
+		FileVersion errorHandlerFileVersion = registerFile(function.getErrorHandlerFile(), ScriptHandler.ERROR_HANDLER_FILE, props, true, executionContext);
 
 		if (configuration.getPropertyAsBoolean("plugins.java.validate.properties")) {
 			props.put(KeywordExecutor.VALIDATE_PROPERTIES, Boolean.TRUE.toString());
 		}
 
-		return props;
+		return new HandlerProperties(props) {

```

Comment:
Alternatively we could also pass all the closeable to the constructor of the HandlerProperties and implement a generic close mechanism